source/material_model/simple_nonlinear.cc

```diff
@@ -0,0 +1,371 @@
+          prm.declare_entry ("Heat capacity", "1.25e3", Patterns::List(Patterns::Double(0)), "Units: $J / (K * kg)$");
+          prm.declare_entry ("Densities", "3300.",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of densities, $\\rho$, for background mantle and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "If only one values is given, then all use the same value.  Units: $kg / m^3$");
+          prm.declare_entry ("Thermal expansivities", "3.5e-5",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of thermal expansivities for background mantle and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "If only one values is given, then all use the same value.  Units: $1 / K$");
+
+
+          // SimpleNonlinear creep parameters
+          prm.declare_entry ("Prefactor", "1e-37",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of viscosity prefactors, $A$, for background mantle and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "If only one values is given, then all use the same value. "
+                             "Units: $Pa^{-n_{dislocation}} m^{n_{dislocation}/m_{dislocation}} s^{-1}$");
+          prm.declare_entry ("Stress exponent", "3",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of stress exponents, $n_dislocation$, for background mantle and compositional fields, "
+                             "for a total of N+1 values, where N is the number of compositional fields. "
+                             "If only one values is given, then all use the same value.  Units: None");
+
+          // averaging parameters
+          prm.declare_entry ("Viscosity averaging p", "-1",
+                             Patterns::Double(),
+                             "This is the p value in the generalized weighed average eqation: "
+                             " mean = \\frac{1}{k}(\\sum_{i=1}^k \\big(c_i \\eta_{\\text{eff}_i}^p)\\big)^{\\frac{1}{p}}. "
+                             " Units: $Pa s$");
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+
+
+    template <int dim>
+    void
+    SimpleNonlinear<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      // Can't use this->n_compositional_fields(), because some
+      // tests never initialize the simulator, but uses the material
+      // model directly.
+      prm.enter_subsection ("Compositional fields");
+      {
+        n_fields = prm.get_integer ("Number of fields")+1;
```

Comment:
Because this material mode was set up based on a material model which worked with a background compositional field. But we still need to store values for that field